src/consensus/network_event.rs

```diff
@@ -217,41 +217,24 @@ impl Debug for NetworkEvent {
     }
 }
 
-// Change to section elders.
+// Neighbour section elders that got removed/demoted.
 #[derive(Default, Eq, PartialEq, Serialize, Deserialize)]
-pub struct EldersChange {
-    // Neighbour peers that became elders.
-    pub neighbours_added: BTreeSet<P2pNode>,
-    // Neighbour peers that ceased to be elders.
-    pub neighbours_removed: BTreeSet<P2pNode>,
-}
+pub struct NeighbourEldersRemoved(pub BTreeSet<P2pNode>);
```

Comment:
OK, cool. I like this one. And now we are getting into the really interesting discussions as we shift towards event-centric code; behaviour.

So previously we had `EldersChange` with added and removed Elders.
Generally I would not consider it a good event, because it is so broad, events should be as granular as possible, but no more than useful. It all depends on the behaviour.
If an Elder is always replaced when removed, then they do come in pairs (removed + added). 
In our case I think it's not very common that there are many of them at a time though? So a whole set of added and removed? Seems there's some serious turmoil for that to be ever used. Normally the Elder changes are so far between, that it would be a very delayed event if we were to wait for subsequent removals.
On the other hand, when they do come in bunches, there might be stress on the network, so I guess the thinking is that at that time it is better to lower stress by not sending more events than necessary.
And as a final thought on all that, there's not really much lost in having a set in the event. I've often designed events like that as well, even when it is most often just ever going to be one single item in it.

Moving onwards: But, Elders aren't always replaced when removed, are they? And even if they are, is it so immediate that we should hold off on the removed event just to await the added one, and bundle them in one event?
That's a good reason to split `NeighbourEldersChanged` to `NeighbourEldersRemoved` and `NeighbourEldersAdded`.
Performance considerations come to mind. Also: who needs this information and how important is it to get it sooner rather than later? I have used events like the former, more broad one, as well at times.

So, when deciding granularity, there's always the "it depends" playing in.

Having said all that, I'm now just curious to how you were reasoning behind this specific event, what motivates this level of granularity, and what happened to `NeighbourEldersAdded`?

Reply:
Let me clarify what the purpose of this struct really is, which is very narrow: we want to identify peers that are no longer elders because we want to disconnect from them. This is because network connections are limited resource and we don't want to waste them. There is really not much more to it than this. Which is why I'm not entirely convinced that this struct is an event. I think it's more like utility / auxilliiary thing.

Couple of points:
1. There is a kind of inconsistency at the network transport layer in that connection are implicit, but disconnection are explicit. If we somehow made even disconnection implicit, then we wouldn't need this struct at all. 
2. Or if we had a way to list all currently active connections, then we would be able to compare this list with the list of known elders and disconnect the ones we no longer need.
2. Or maybe this whole logic could at least be moved somewhere else? For example, maybe the function that performs the update of the neighbour containers could return the set of no-longer-elders itself?

Reply:
It seems my questions got lost.
So, apparently, you did the change from EldersChange to NeighbourEldersRemoved, my question is what happened to NeighbourEldersAdded?
The previous struct had info on both, now there's only one. 

So, what was the reasoning there? Don't we need to know about the added ones?

About being an event or not, if it described something that happened, then it essentially is an event. If that "something" was already described in an event, then there are numerous considerations: 
- Was this info part of that event? 
- Is this info passed between some subscriber and someone else?
- Why is "someone else" not also a subscriber to the event? (That's the normal way, one event often has multiple subscribers, with various responsibilities.)
If the event is not passed to the subscriber because it has too much info, the event might be having too much responsibility, or there is some key part missing in the call flow between these parts.


There are many unclarities with regards to what is considered with the design and the rationale.
Most often I would say that when we can't clearly describe a pattern and its rationale, it's because there is none.

Reply:
Ah, sorry. There is no `NeighbourEldersAdded` because it's not needed. As said, we only need to track the removed elders so we can disconnect from them. We don't need to do anything with the added ones, because the connections are implicit (will be established on first message sent to them).

Regarding event/not event - there is no subscription going on with this thing. It's produced on one line and then consumed couple of lines later within the same function. But perhaps it's still justified to call it an event? Just one with very narrow scope?

Reply:
I don't think this part is very important, but I will have a closer look at the code when I'm at the machine to make some further comments and/or conclusions.

But just quick question, was the added fields in previous struct not used, or did the use of the struct change?

Reply:
It wasn't used.